### PR TITLE
libckteec: Add support for digest operations

### PR DESCRIPTION
**Pull request by vesajaaskelainen**
Implements support for digest operations as specified in:

PKCS #11 Cryptographic Token Interface Base Specification Version 2.40
Plus Errata 01
5.10 Message digesting functions

Signed-off-by: Vesa Jääskeläinen <vesa.jaaskelainen@vaisala.com>

### CODE REVIEW

**Comment by vesajaaskelainen** [comment]
Updated the PR to match functionality changes in TA PR side.

File: libckteec/src/pkcs11_processing.c [hunk 1]
```diff
@@ -1087,3 +1087,35 @@ CK_RV ck_derive_key(CK_SESSION_HANDLE session,
...
+	if (!ctrl)
+		return CKR_HOST_MEMORY;
+
+	buf = ctrl->buffer;
+
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
+
+	memcpy(buf, &cmd, sizeof(cmd));
+	buf += sizeof(cmd);
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:1393]: not needed

File: libckteec/src/pkcs11_processing.c [hunk 2]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	if (!ctrl)
+		return CKR_HOST_MEMORY;
+
+	buf = ctrl->buffer;
+
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
+
+	memcpy(buf, &key_handle, sizeof(key_handle));
+	buf += sizeof(key_handle);
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:413]: can remove this line

File: libckteec/src/pkcs11_processing.c [hunk 3]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	TEEC_SharedMemory *ctrl = NULL;
+	TEEC_SharedMemory *io1 = NULL;
+	uint32_t session_handle = session;
+	uint8_t *buf = NULL;
+
+	/*
+	 * Bad argument handling for in and in_len:
+	 * - If in == NULL AND in_len != 0 -> need to call TA to
+	 * terminate session.
+	 * - If in == NULL and in_len == 0 -> just operate normally.
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:434]: s/and/AND/

File: libckteec/src/pkcs11_processing.c [hunk 4]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	}
+
+	/*
+	 * Shm io0: (in/out) ctrl = [session-handle] / [status]
+	 */
+	ctrl = ckteec_alloc_shm(sizeof(session_handle), CKTEEC_SHM_INOUT);
+	buf = ctrl->buffer;
+
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:449]: can remove this line

File: libckteec/src/pkcs11_processing.c [hunk 5]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	TEEC_SharedMemory *io2 = NULL;
+	uint32_t session_handle = session;
+	size_t out_size = 0;
+	uint8_t *buf = NULL;
+
+	/*
+	 * Bad argument handling for in and in_len:
+	 * - If in == NULL AND in_len != 0 -> need to call TA to
+	 * terminate session.
+	 * - If in == NULL and in_len == 0 -> just operate normally.
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:485]: ditto

File: libckteec/src/pkcs11_processing.c [hunk 6]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+
+	/*
+	 * Bad argument handling for digest_ref and digest_len:
+	 * - If digest_ref == NULL AND digest_len == NULL -> need to call TA to
+	 * terminate session.
+	 * - If digest_len == NULL -> need to call to TA to terminate session.
+	 * - If digest_ref == NULL BUT digest_len != NULL just operate normally
+	 * to get size of the required buffer
+	 */
+	if ((!digest_ref && !digest_len) || !digest_len) {
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:501]: in the end: `if (!digest_len) {`

File: libckteec/src/pkcs11_processing.c [hunk 7]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	 */
+	ctrl = ckteec_alloc_shm(sizeof(session_handle), CKTEEC_SHM_INOUT);
+	if (!ctrl) {
+		rv = CKR_HOST_MEMORY;
+		goto bail;
+	}
+	buf = ctrl->buffer;
+
+	memcpy(buf, &session_handle, sizeof(session_handle));
+	buf += sizeof(session_handle);
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:518]: can remove this line

File: libckteec/src/pkcs11_processing.c [hunk 8]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	if (in_len && in) {
+		io1 = ckteec_register_shm(in, in_len, CKTEEC_SHM_IN);
+		if (!io1) {
+			rv = CKR_HOST_MEMORY;
+			goto bail;
+		}
+	}
+
+	/* Shm io2: output digest or null sized shm */
+	if (digest_len && digest_ref) {
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:530]: `	if (digest_ref) {`
The tests above ensure `digest_len` != NULL.

File: libckteec/src/pkcs11_processing.c [hunk 9]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+
+	if (!io2) {
+		rv = CKR_HOST_MEMORY;
+		goto bail;
+	}
+
+	rv = ckteec_invoke_ta(PKCS11_CMD_DIGEST_ONESHOT, ctrl,
+			      io1, io2, &out_size, NULL, NULL);
+
+	if (digest_len && (rv == CKR_OK || rv == CKR_BUFFER_TOO_SMALL))
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:546]: here also, no need to test `digest_len`

File: libckteec/src/pkcs11_processing.c [hunk 10]
```diff
@@ -341,6 +341,285 @@ CK_RV ck_encdecrypt_final(CK_SESSION_HANDLE session,
...
+	uint8_t *buf = NULL;
+
+	/*
+	 * - If digest_ref == NULL AND digest_len == NULL -> need to call TA to
+	 * terminate session.
+	 * - If digest_len == NULL -> need to call to TA to terminate session.
+	 * - If digest_ref == NULL BUT digest_len != NULL just operate normally
+	 * to get size of the required buffer
+	 */
+	if ((!digest_ref && !digest_len) || !digest_len) {
```
Comment (etienne-lms) [inline libckteec/src/pkcs11_processing.c:577]: ditto

**Review by etienne-lms — COMMENTED** [review]
For commit "libckteec: Add API for releasing active processing":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` with a minor comment.

Few comments for commit "libckteec: Add support for digest operations".

File: libckteec/src/pkcs11_processing.c (hunk 2 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:413]: OK

File: libckteec/src/pkcs11_processing.c (hunk 3 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:434]: OK

File: libckteec/src/pkcs11_processing.c (hunk 4 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:449]: OK

File: libckteec/src/pkcs11_processing.c (hunk 5 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:485]: OK

File: libckteec/src/pkcs11_processing.c (hunk 7 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:518]: OK

File: libckteec/src/pkcs11_processing.c (hunk 1 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:1393]: OK

File: libckteec/src/pkcs11_processing.c (hunk 6 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:501]: OK

File: libckteec/src/pkcs11_processing.c (hunk 8 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:530]: OK

File: libckteec/src/pkcs11_processing.c (hunk 9 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:546]: OK

File: libckteec/src/pkcs11_processing.c (hunk 10 above)
Comment (vesajaaskelainen) [reply libckteec/src/pkcs11_processing.c:577]: OK

**Comment by vesajaaskelainen** [comment]
Whups -- slightly more commits than was originally intended ;)

**Comment by vesajaaskelainen** [comment]
Updated the PR with the review feedback.

Applied tag for `libckteec: Add API for releasing active processing`.

File: libckteec/include/pkcs11_ta.h [hunk 11]
```diff
@@ -581,6 +581,19 @@ enum pkcs11_ta_cmd {
 	 * This command relates to the PKCS#11 API function C_DeriveKey().
 	 */
 	PKCS11_CMD_DERIVE_KEY = 43,
+
+	/*
+	 * PKCS11_CMD_RELEASE_ACTIVE_PROCESSING - Release active processing
+	 *
+	 * [in]  memref[0] = 32bit session handle, 32bit enum pkcs11_ta_cmd
```
Comment (etienne-lms) [inline libckteec/include/pkcs11_ta.h:588]: sorry, i forgot these. For consistency with other command description, prefer split this in several lines:
```
	 * [in]  memref[0] = [
	 *              32bit session handle,
	 *              32bit enum pkcs11_ta_cmd
	 *       ]

Comment (vesajaaskelainen) [reply libckteec/include/pkcs11_ta.h:588]: OK

**Review by etienne-lms — COMMENTED** [review]
For commit "libckteec: Add support for digest operations":
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by ruchi393** [comment]
For the PR:

`Reviewed-by: Ruchika Gupta <ruchika.gupta@linaro.org>`

File: libckteec/include/pkcs11_ta.h (hunk 11 above)
Comment (vesajaaskelainen) [reply libckteec/include/pkcs11_ta.h]: Will update the TA PR with this too.

**Pull request merged**